### PR TITLE
Fixes #2667 Retry decorator bug and unit tests

### LINKED ISSUE #2667: Fix bug in retry decorator and add tests
Code here:
https://github.com/numenta/nupic/blob/master/src/nupic/support/decorators.py

See PR for numenta-apps fix to similar code here:
https://github.com/numenta/numenta-apps/pull/399

See PR for the similar tests in numenta-apps code here:
https://github.com/numenta/numenta-apps/pull/382 (it's work-in-progress; wait for it to be completed/merged before borrowing the test code)

@saganbolliger @vitaly-krugl

### PR DESCRIPTION
**Pull request by SaganBolliger**
Fixes #2667 

Brings over retry decorator bug fix from numenta/numenta-apps#399
Brings over unit tests for retry decorator from numenta/numenta-apps#382

@scottpurdy @vitaly-krugl

### CODE REVIEW

**Comment by scottpurdy** [comment]
@SaganBolliger - One of the Travis jobs timed out so I restarted it.

@vitaly-krugl - Please review.

**Comment by SaganBolliger** [comment]
@scottpurdy Looks like Travis timed out again.

**Comment by vitaly-krugl** [comment]
Travis was having trouble installing MySQL. CC @rhyolight 
```
==> Downloading https://downloads.sf.net/project/machomebrew/Bottles/mysql-5.6.2
#####################                                                     30.1%
curl: (56) Recv failure: Operation timed out
Error: Failed to download resource "mysql"
Download failed: https://downloads.sf.net/project/machomebrew/Bottles/mysql-5.6.21.mavericks.bottle.tar.gz
Warning: Bottle installation failed: building from source.
==> Downloading http://cdn.mysql.com/Downloads/MySQL-5.6/mysql-5.6.21.tar.gz
curl: (22) The requested URL returned error: 404 Not Found
Error: Failed to download resource "mysql"
Download failed: http://cdn.mysql.com/Downloads/MySQL-5.6/mysql-5.6.21.tar.gz
>>> mysql.server start
./ci/travis/before_install-osx.sh: line 59: mysql.server: command not found
The command "if [ ${TRAVIS_OS_NAME:-'linux'} = 'osx' ]; then . ./ci/travis/before_install-osx.sh; fi" failed and exited with 127 during .
Your build has been stopped.
```

File: tests/unit/nupic/support/decorators_test.py [hunk 1]
```diff
@@ -22,18 +22,251 @@
...
+from mock import patch, call, Mock
 
-from mock import patch, Mock
+from nupic.support.unittesthelpers.testcasebase import unittest
+from nupic.support import decorators
 
 
-from nupic.support.unittesthelpers.testcasebase import unittest
 
+class TestParentException(Exception):
```
Comment (vitaly-krugl) [inline tests/unit/nupic/support/decorators_test.py]: There should be 3 lines between global-level constructs (e.g., class definitions, functions)

File: tests/unit/nupic/support/decorators_test.py [hunk 2]
```diff
@@ -22,18 +22,251 @@
...
+  pass
```
Comment (vitaly-krugl) [inline tests/unit/nupic/support/decorators_test.py]: There should be 3 lines between global-level constructs (e.g., class definitions, functions)

File: tests/unit/nupic/support/decorators_test.py [hunk 3]
```diff
@@ -22,18 +22,251 @@
...
+
+
+  @patch("time.sleep", autospec=True)
+  @patch("time.time", autospec=True)
+  def testRetryWaitsInitialRetryDelaySec(self, mockTime, mockSleep):
+    """Test that delay times are correct."""
+
+    self.mockSleepTime(mockTime, mockSleep)
+
+    _retry = decorators.retry(
```
Comment (vitaly-krugl) [inline tests/unit/nupic/support/decorators_test.py]: A leading underscore usually indicates a private object reference in python. I don't think it's appropriately used here (and in the other new test methods here), and it impacts readability of the code. I must have missed this during the review of the same code in the numenta-apps repo.

`retryDecorator` would be a good alternative.

File: tests/unit/nupic/support/decorators_test.py [hunk 4]
```diff
@@ -22,18 +22,251 @@
...
+  @patch("time.time", autospec=True)
+  def testRetryRetryFilter(self, mockTime, mockSleep):
+    """Test that if retryFilter is specified and exception is in
+    retryExceptions, retries iff retryFilter returns true."""
+
+    self.mockSleepTime(mockTime, mockSleep)
+
+    # Test with retryFilter returning True
+
+    _retryTrueFilter = decorators.retry(
```
Comment (vitaly-krugl) [inline tests/unit/nupic/support/decorators_test.py]: Shouldn't be using the underscore in the name `_retryTrueFilter`. I see more of this in other local variables. Please fix them all.

**Comment by vitaly-krugl** [comment]
@SaganBolliger: I am done with the review and there is feedback. Please also make sure that the changes pass cleanly through pylint using nupic's pylintrc.

**Comment by SaganBolliger** [comment]
@vitaly-krugl I've incorporated your feedback and cleaned up a few lint issues. I believe it's ready to merge.

**Comment by vitaly-krugl** [comment]
@SaganBolliger: Just as an FYI, nupic uses the Doxygen confention, while numenta-apps repo uses sphinx restructuredText format (http://sphinx-doc.org/domains.html#python-signatures)

File: tests/unit/nupic/support/decorators_test.py [hunk 5]
```diff
@@ -22,110 +22,345 @@
...
+    with self.assertRaises(TestChildException):
+      testFunction()
+
+    self.assertEqual(mockSleep.call_count, 1)
+
+
+  @patch("time.sleep", autospec=True)
+  @patch("time.time", autospec=True)
+  def testRetryRetryExceptionExcluded(self, mockTime, mockSleep):
+    """ Test that retry is not triggered if raised exeception is not in
```
Comment (vitaly-krugl) [inline tests/unit/nupic/support/decorators_test.py]: spelling "exeception"

File: tests/unit/nupic/support/decorators_test.py [hunk 6]
```diff
@@ -22,110 +22,345 @@
...
+        TestParentException("Test exception 1"),
+        TestParentException("Test exception 2"),
+        321
+      ],
+      __name__="testFunction", autospec=True)
+
+    returnValue = retryDecorator(testFunction)()
+
+    self.assertEqual(returnValue, 321)
+    self.assertEqual(testFunction.call_count, 3)
```
Comment (vitaly-krugl) [inline tests/unit/nupic/support/decorators_test.py]: Just wanted to remind about adding test(s) for the logging arg to the retry decorator. Both passed and not passed.

**Comment by vitaly-krugl** [comment]
@SaganBolliger: I am done with the review. Okay to merge after fixing the noted spelling mistake. I think that you were planning to add test(s) of the log arg in a separate PR.

**Comment by SaganBolliger** [comment]
@vitaly-krugl Spelling mistake fixed. Yes, will add tests for logging in separate PR.

**Comment by SaganBolliger** [comment]
@vitaly-krugl I've squashed all the commits. Should be ready for merge.

**Comment by vitaly-krugl** [comment]
Thanks @SaganBolliger !

**Pull request merged**